source/common/ur_singleton.h

```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ * Copyright (C) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+#pragma once
+#include <memory>
+#include <unordered_map>
+#include <mutex>
+
+//////////////////////////////////////////////////////////////////////////
+/// a abstract factory for creation of singleton objects
+template<typename _singleton_t, typename _key_t>
+class singleton_factory_t
+{
+protected:
+    using singleton_t = _singleton_t;
+    using key_t = typename std::conditional<std::is_pointer<_key_t>::value, size_t, _key_t>::type;
+
+    using ptr_t = std::unique_ptr < singleton_t >;
+    using map_t = std::unordered_map < key_t, ptr_t >;
+
+    std::mutex mut; ///< lock for thread-safety
+    map_t map;      ///< single instance of singleton for each unique key
+
+    //////////////////////////////////////////////////////////////////////////
+    /// extract the key from parameter list and if necessary, convert type
+    template<typename... Ts>
+    key_t getKey( _key_t _key, Ts&&... _params )
+    {
+        return reinterpret_cast<key_t>( _key );
+    }
+
+public:
+    //////////////////////////////////////////////////////////////////////////
+    /// default ctor/dtor
+    singleton_factory_t() = default;
+    ~singleton_factory_t() = default;
+
+    //////////////////////////////////////////////////////////////////////////
+    /// gets a pointer to a unique instance of singleton
+    /// if no instance exists, then creates a new instance
+    /// the params are forwarded to the ctor of the singleton
+    /// the first parameter must be the unique identifier of the instance
+    template<typename... Ts>
+    _singleton_t* getInstance( Ts&&... _params )
+    {
+        auto key = getKey( std::forward<Ts>( _params )... );
+
+        if(key == 0) // No zero keys allowed in map
+            return static_cast<_singleton_t*>(0);
```

Comment:
My preference would be to always have `{}` in any control flow statements.

Reply:
That's my preference as well. But I refrained from making any stylistic changes (the coding style is all over the place :-)). I suggest we pick on a single style and enforce it with clang-format (we were planning on doing that soon after this PR was merged).

I can do it with this PR, but a) it might conflict with the generated files, so it's not trivial addition, and b) it would probably affect the entire repo, not just the loader, so might clutter the PR.

Reply:
Done. I'll add an issue about adding style checking to the build system.

Reply:
#101

Reply:
Addressing them in a future issue works for me :+1: